tests/TestCase/AuthenticationServiceTest.php

```diff
@@ -104,6 +103,29 @@ public function testAuthenticateWithChallenge()
         }
     }
 
+    public function testAuthenticateHttpAuthCredentialsPresent()
+    {
+        $request = ServerRequestFactory::fromGlobals([
+            'SERVER_NAME' => 'example.com',
+            'REQUEST_URI' => '/testpath',
+            'PHP_AUTH_USER' => 'mariano',
+            'PHP_AUTH_PW' => 'WRONG',
```

Comment:
It might be good to check that not sending a challenge still lets the user re-enter their credentials in a browser when they make a mistake.

Reply:
My mistake, I had overlooked this case. Pretty sure the browsing won't show the dialog to re-enter the credentials without the `www-authenticate` header. I checking using HTTP Basic Auth through Apache config and it too returns a 401 with the `www-authenticate` header in case of invalid credentials.

So I have reverted the changes I had done in `AuthenticationService`. Just updated `HttpBasicAuthenticator` to return `Result` with proper status when credentials are missing vs. incorrect credentials.

Should I add a config to the authenticator to allow skipping the challenge in case of missing/incorrect credentials?

Reply:
> Should I add a config to the authenticator to allow skipping the challenge in case of missing/incorrect credentials?

That sounds good to me. I don't think not emitting the challenge justifies another adapter.

Reply:
Done